src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs

```diff
@@ -327,9 +328,15 @@ public override bool HandleCall(MethodIL callingMethodBody, MethodDesc calledMet
                                         marshalAsDescriptor = paramMetadata[metadataIndex].MarshalAsDescriptor;
                                 }
 
+                                aotUnsafeDelegate |= IsAotUnsafeDelegate(calledMethod.Signature[paramIndex]);
                                 comDangerousMethod |= IsComInterop(marshalAsDescriptor, calledMethod.Signature[paramIndex]);
                             }
 
+                            if (aotUnsafeDelegate)
+                            {
+                                diagnosticContext.AddDiagnostic(DiagnosticId.CorrectnessOfAbstractDelegatesCannotBeGuaranteed, calledMethod.GetDisplayName());
```

Comment:
Before this PR, we got warning for AOT unsafe delegates in structs, like:
```
foo(default);

[DllImport("bla")]
static extern void foo(MyStruct s);

struct MyStruct
{
    Delegate del;
}
```
Is this replacement still going to issue a warning for this case?

Reply:
These warnings are best effort right now. If we want to patch up the holes, we'll need to declare a lot of interop as AOT unsafe (AsAny and LayoutClass marshalling can also put is into this situation at minimum and I'm sure I'm missing more because the original architects of .net interop were maximalists). If we patch those up, we can add a check here (and no longer need to worry about hitting the recursion in the new spot).

Reply:
I see that there are number of holes today. I think we should generate warnings for all AOT unsafe interop. I have opened https://github.com/dotnet/runtime/issues/74697 on it.